File: LeetCode/496.NextGreater_Element_I/susan/496.NextGreater_Element_I.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] res1 = new int[nums1.length];
        HashMap<Integer,Integer> map = new HashMap<>();
        Stack<Integer> s = new Stack<>();

        for(int i = nums2.length - 1; i>= 0; i--){
           while(!s.isEmpty() && s.peek() <= nums2[i]){
               s.pop();
           }
           int res = s.isEmpty()? -1:s.peek();
           map.put(nums2[i],res);
           s.push(nums2[i]);
        }

        for(int j = 0; j< nums1.length ; j++){
            if(map.containsKey(nums1[j])){
                res1[j] = map.get(nums1[j]);
            }
        }

        return res1;

    }
}

File: LeetCode/496.Next_Greater_Element_I/Mario/496.Next_Greater_Element_I.py
#!/usr/bin/env python3


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # get nextGreaterElement for nums2
        # stack = [] store index,
        # if hit num which <= nums2[stack[-1]], stack.append(i)
        # else while

        greater = {}
        stack = []

        for num in nums2:
            while stack and stack[-1] < num:
                last = stack.pop()
                greater[last] = num

            stack.append(num)

        # print(f"greater: {greater}")

        res = []
        for i, num in enumerate(nums1):
            val = -1 if num not in greater else greater[num]
            res.append(val)

        return res
